lib/cache/inventory_cache.go

```diff
@@ -0,0 +1,655 @@
+		return nil, trace.Wrap(err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	ic := &InventoryCache{
+		cfg: cfg,
+
+		// Create the sortcache
+		store: newStore(
+			"unified_instance",
+			func(u *inventoryInstance) *inventoryInstance {
+				return u.clone()
+			},
+			map[inventoryIndex]func(*inventoryInstance) string{
+				inventoryAlphabeticalIndex: func(u *inventoryInstance) string {
+					return u.getAlphabeticalKey()
+				},
+				inventoryTypeIndex: func(u *inventoryInstance) string {
+					return u.getTypeKey()
+				},
+				inventoryIDIndex: func(u *inventoryInstance) string {
+					return u.getIDKey()
+				},
+			},
+		),
+
+		// Create a channel that will close when the initialization is done.
+		initDone: make(chan struct{}),
+
+		ctx:    ctx,
+		cancel: cancel,
+	}
+
+	go ic.initializeAndWatchWithRetry(ctx)
+
+	return ic, nil
+}
+
+// IsHealthy returns true if the cache is healthy and initialized.
+func (ic *InventoryCache) IsHealthy() bool {
+	return ic.healthy.Load()
+}
+
+func (ic *InventoryCache) Close() error {
+	ic.cancel()
+	// Wait for initDone channel to finish so we can close gracefully.
+	<-ic.initDone
+	return nil
+}
+
+// calculateReadsPerSecond calculates the rate limit to use for backend reads based on cluster size.
+// With this implementation, these are some of the expected rate limits and corresponding total times based on cluster size:
+//
+// Cluster size | Reads per second | Total time to finish all reads
+// -------------|------------------|-------------------------------
+// 500          | 283              | 1.77s
+// 1,000        | 298              | 3.36s
+// 2,000        | 322              | 6.21s
+// 4,000        | 363              | 11.02s
+// 8,000        | 433              | 18.47s
+// 32,000       | 789              | 40.56s
+// 64,000       | 1219             | 52.5s
+// 128,000      | 2035             | 1m03s
+// 256,000      | 3605             | 1m11s
+func calculateReadsPerSecond(clusterSize int) int {
+	// minimumComponent is the minimum value of reads per second we never want to drop below.
+	const minimumComponent = 256
+
+	// linearComponent ensures we stay under a worst-case upper bound init time of 90s.
+	linearComponent := clusterSize / 90
+
+	// subLinearComponent ensures that growth is sub-linear across most reasonable cluster sizes.
+	subLinearComponent := int(math.Sqrt(float64(clusterSize)))
+
+	return minimumComponent + linearComponent + subLinearComponent
+}
+
+// initializeAndWatchWithRetry runs initializeAndWatch with a retry every 10 seconds if it fails.
+func (ic *InventoryCache) initializeAndWatchWithRetry(ctx context.Context) {
+	defer close(ic.initDone)
+
+	const retryInterval = 10 * time.Second
+
+	for {
+		ic.cfg.Logger.DebugContext(ctx, "Attempting to initialize inventory cache")
+
+		// Attempt to initialize and watch
+		if err := ic.initializeAndWatch(ctx); err != nil {
+			if ctx.Err() != nil {
+				return
```

Comment:
Something like this?

```suggestion
			if ctx.Err() != nil {
				ic.cfg.Logger.DebugContext(ctx, "Exiting from inventory cache watch loop due to context cancellation")
				return
```